### ISSUE TITLE
MimeKit.ParseException: Invalid local-part at offset 0 on Gmail Imap

### ISSUE THREAD

**Issue by mbalsam** [jstedfast/MailKit#494]
This just started getting an exception being thrown from inside of Imap.ImapFolder.Fetch() with my gmail account.    See below.  

I understand that local-part refers to the part to the left of the @ in the email address.  With the Imap diagnostic turned on, i found a French language messages a few messages before the location with the stack trace.  It that has this in the local-part:

`From: =?UTF-8?Q?Lib=C3=A9ration?= <newsletter@liberation.cccampaigns.net>`

The gmail "show original" shows it as 

`From: "Libération" <newsletter@liberation.cccampaigns.net>`

Disclaimer:  I'm guessing that this is the messages that's causing the crash.  I cant be sure since the parse and download streams appear to be on different threads.  

This is the full stack trace: 
```
MimeKit.ParseException: Invalid local-part at offset 0
   at MimeKit.InternetAddress.TryParseLocalPart(Byte[] text, Int32& index, Int32 endIndex, Boolean throwOnError, String& localpart)
   at MimeKit.InternetAddress.TryParseAddrspec(Byte[] text, Int32& index, Int32 endIndex, Byte[] sentinels, Boolean throwOnError, String& addrspec, Int32& at)
   at MimeKit.MailboxAddress.set_Address(String value)
   at MimeKit.MailboxAddress..ctor(Encoding encoding, String name, String address)
   at MailKit.Net.Imap.ImapUtils.EnvelopeAddress.ToMailboxAddress()
   at MailKit.Net.Imap.ImapUtils.ParseEnvelopeAddressList(InternetAddressList list, ImapEngine engine, String format, CancellationToken cancellationToken)
   at MailKit.Net.Imap.ImapUtils.ParseEnvelope(ImapEngine engine, CancellationToken cancellationToken)
   at MailKit.Net.Imap.ImapFolder.FetchSummaryItems(ImapEngine engine, ImapCommand ic, Int32 index)
   at MailKit.Net.Imap.ImapEngine.ProcessUntaggedResponse(CancellationToken cancellationToken)
   at MailKit.Net.Imap.ImapCommand.Step()
   at MailKit.Net.Imap.ImapEngine.Iterate()
   at MailKit.Net.Imap.ImapEngine.Wait(ImapCommand ic)
   at MailKit.Net.Imap.ImapFolder.Fetch(IList`1 uids, MessageSummaryItems items, HashSet`1 fields, CancellationToken cancellationToken)
```

**Comment by jstedfast** [comment]
I'll need a protocol log (see the FAQ.md file on how to get one) because it's not parsing the raw message, it's parsing the pre-parsed address that the IMAP server is giving MailKit.

**Comment by jstedfast** [comment]
Also, I guarantee that's not the problematic email address. It has a perfectly valid local-part.

**Comment by mbalsam** [comment]
I've generated the protocol log and zipped it..  I think its too big for github  315K??   I will send it to your microsoft email account. 

Many thanks for looking into this.

**Comment by jstedfast** [comment]
Okay, it's fixed now.

**Comment by mbalsam** [comment]
Works. many thanks!

**Comment by jstedfast** [comment]
No problem :)

**Comment by mbalsam** [comment]
Think i found another variation. 

		address	"=?UTF-8?Q?\"N=C3=BA=C3=B1ez_y_Navarro\"?="	string
		name	"<comunicacion=orbyce.com@mail39.atl71.mcdlv.net>"	string

[imap-4-19-1-01.txt](https://github.com/jstedfast/MailKit/files/936358/imap-4-19-1-01.txt)

**Comment by mbalsam** [comment]
In this one it appears that the address and name are reversed?

**Comment by jstedfast** [comment]
lol, yes, that is what it looks like...

**Comment by jstedfast** [comment]
I meant to get to this tonight, but spent the night debugging YDecoder in MimeKit instead :-\

I'll fix this up in the morning.

**Comment by mbalsam** [comment]
Works.   Thanks for the quick turnaround!!!

**Comment by sivajallepalli** [comment]
Still I'm facing below issue with version 1.14.0 and higher versions when I use email address as "xyz <<xyz@test.com>>"

`MimeKit.ParseException: Invalid addr-spec token at offset 0
   at MimeKit.InternetAddress.TryParseAddrspec(Byte[] text, Int32& index, Int32 endIndex, Byte[] sentinels, Boolean throwOnError, String& addrspec, Int32& at)
   at MimeKit.MailboxAddress.set_Address(String value)
`

**Comment by jstedfast** [comment]
The MailboxAddress .ctor is not meant to parse anything. Use MailboxAddress.Parse() instead.

**Comment by ooples** [comment]
I'm getting this error as well and I'm initializing a new MailBoxAddress item so should I be using the MailBoxAddress.Parse() method instead?

**Comment by jstedfast** [comment]
@ooples yes.

The .ctor is not meant to parse arbitrary text into a DisplayName / Address pair.

**Comment by ooples** [comment]
@jstedfast Are there examples for the Parse method? I'm seeing a bunch of overloads and I'm just trying to change something simple like this: 

toList.Add(new MailBoxAddress(email, name)) into toList.Add(MailBoxAddress.Parse(?))

**Comment by jstedfast** [comment]
@ooples why not just do this if you already have the individual display-name/address strings?

```csharp
toList.Add (new MailboxAddress (name, email));
```

The Parse() method expects to find the name in the input string. That's why it doesn't have an option to pass in the name.

**Comment by ooples** [comment]
@jstedfast I don't know if you saw the pseudo code in my post up above but that is exactly what I'm doing now and I'm getting the same exception. I figured the only way to fix the issue without everything breaking would be to use the Parse or TryParse methods but since I couldn't find any examples anywhere online I just wrote my own rudimentary code. Tbh I don't even know why I'm getting this exception in the first place because I double checked the emails and they are all valid.
